_pytest/main.py

```diff
@@ -476,9 +488,39 @@ def _parsearg(self, arg):
                     " (missing __init__.py?)")
             else:
                 raise UsageError("file not found: " + arg)
+        if self.line:
+            self.nearest_func_name = self._get_nearest_func_name(self.line, path)
         parts[0] = path
         return parts
 
+    def _get_nearest_func_name(self, line, path):
+        path = str(path)
+        start_line = line = int(line) - 1
+        if not os.path.isfile(path):
+            raise UsageError("{} is not a file".format(path))
+        with open(path) as f:
+            content = f.readlines()
+        at_flag = False
+        pattern_func = r'[\t ]*?def test(?P<func_name>[a-zA-Z0-9_-]+)\(*'
```

Comment:
This regex won't catch functions using non-ascii names. I think we can use `str.isidentifier` instead for that.

Reply:
```
In [18]: 'def test_01():'.isidentifier()
Out[18]: False
```
it returns False :(

Reply:
Yeah I meant the second case (combining the regex and `str.isidentifier`), and indeed it works for non-ascii names (I did not know that), but it won't work for non-character names like `def testλ():` (which is valid):

```python
>>> import re
>>> def testλ(): pass
...
>>> re.match(r'[\t ]*?def test(?P<func_name>[a-zA-Z0-9_-]+)\(*', 'def testλ():')
>>> re.match(r'[\t ]*?def test(?P<func_name>[a-zA-Z0-9_-]+)\(*', 'def test_не_англ_имя_01():')
<_sre.SRE_Match object; span=(0, 9), match='def test_'>
```

Reply:
https://regex101.com/r/0kZAJV/2

It works well with `def testλ(): pass` :)

Reply:
Sorry to be "that guy", but it seems to me that re-reading the file and trying to parse Python with regexes is entirely the wrong approach here.

We already have the function objects collected from this file available somewhere, right? If so, why can't we just use something like [inspect.getsourcelines()](https://docs.python.org/3/library/inspect.html#inspect.getsourcelines) on those?

Reply:
@The-Compiler 
this function wants a code object
```
local('/home/username/tests/test_1.py') is not a module, class, method, function, traceback, frame, or code object
```
Could you advice how I can get this object from a path? Or where this object generates in pytest?
```
_pytest._code.compile(path)
``` 
is not working with path too.

Reply:
@The-Compiler you are not alone, I'm more than uneasy about the current approach.

I just realized something though, we should just store the line number given by the user and *then* filter the items in `pytest_collection_modifyitems` because each `item` already has a `location` attribute which returns a tuple of `(path, linenum, name)`. This will make things *much*
 simpler. 😁

Reply:
@nicoddemus I like it :) But you may have decorators. for example:

```
6 @pytest.mark.parametrize('x', [1,2])
7 def test_02(x):
8     assert 1
```
User run command `pytest test_1.py:6`
I filter it and get item.location:
```
test_01 -> 3
test_02 -> 7
```
And I need to run test on 7th string, but if I don't have decorator I need to run test on 3th string because it may look like 
```
3 def test_01():
4     assert 1
5
6
7 def test_02():
8     assert 1
```
How I can understand it in this case?

Reply:
@nicoddemus I mean this situation:
```
1 import pytest
2
3
4 def test_01():
5     assert 1
6
7
8 @pytest.mark.parametrize('x', [1,2,)
9 def test_02(x):
10     assert 1
11
```

I forgot `]` in the decorator. Pytest writes my a Syntax Error:
```
    __import__(modname)
E     File "/home/username/pytest_3224/tests/test_1.py", line 8
E       @pytest.mark.parametrize('x', [1,2,)
E                                          ^
E   SyntaxError: invalid syntax
```

User might want to run test with line 8. He will not copy/paste until line, but he can write line in the end.

It's just an example, but If it doesn't matter then ok :)

Reply:
@nicoddemus Fixed, but, could you help with algorithm? May you can advice how to do it more beautiful?